### PR TITLE
[FrameworkBundle] Debug container environment variables

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #...   
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

This PR is an attempt to ease debugging environment variables used by the container. Main purpose is to tell which are missing and show detailed usage information.

image

image

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php [hunk 1]
```diff
@@ -199,12 +206,15 @@ protected function getContainerBuilder()
         if (!$kernel->isDebug() || !(new ConfigCache($kernel->getContainer()->getParameter('debug.container.dump'), true))->isFresh()) {
             $buildContainer = \Closure::bind(function () { return $this->buildContainer(); }, $kernel, get_class($kernel));
             $container = $buildContainer();
-            $container->getCompilerPassConfig()->setRemovingPasses(array());
-            $container->compile();
         } else {
             (new XmlFileLoader($container = new ContainerBuilder(), new FileLocator()))->load($kernel->getContainer()->getParameter('debug.container.dump'));
+            $container->setParameter('container.build_hash', $hash = ContainerBuilder::hash(__METHOD__));
+            $container->setParameter('container.build_id', hash('crc32', $hash.time()));
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]: why these changes ?

Comment (ro0NL) [reply src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php]:
```
The service "cache.adapter.apcu" has a dependency on a non-existent parameter "container.build_id".

The service "security.authentication.listener.anonymous.main" has a dependency on a non-existent parameter "container.build_hash".
```

should they be dumped instead? cc @nicolas-grekas

**Comment by ro0NL** [comment]
screenshot updated and JSON version included. I think the data set is pretty nice, and leaves us figuring out how to display processors/processed value in the text table without cluttering output too much. Still need to extract the actual prefixes though :)

**Review by fabpot — COMMENTED** [review]
I really like this one and I would like to get it for 4.2. @ro0NL Can you finish it? Any comments @nicolas-grekas

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/MarkdownDescriptor.php [hunk 2]
```diff
@@ -263,6 +263,14 @@ protected function describeContainerParameter($parameter, array $options = array
         $this->write(isset($options['parameter']) ? sprintf("%s\n%s\n\n%s", $options['parameter'], str_repeat('=', strlen($options['parameter'])), $this->formatParameter($parameter)) : $parameter);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    protected function describeContainerEnvVars(array $envs, array $options = array())
+    {
+        // @TODO
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/MarkdownDescriptor.php]: TODO :)

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php [hunk 3]
```diff
@@ -24,6 +24,8 @@
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 use Symfony\Component\Routing\Route;
 use Symfony\Component\Routing\RouteCollection;
+use Symfony\Component\VarDumper\Cloner\VarCloner;
+use Symfony\Component\VarDumper\Dumper\CliDumper;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: var-dumper is not a mandatory dep of fwb, can we do without?

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/XmlDescriptor.php [hunk 4]
```diff
@@ -130,6 +130,14 @@ protected function describeContainerParameter($parameter, array $options = array
         $this->writeDocument($this->getContainerParameterDocument($parameter, $options));
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    protected function describeContainerEnvVars(array $envs, array $options = array())
+    {
+        // @TODO
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/XmlDescriptor.php]: TODO :)

**Review by nicolas-grekas — COMMENTED** [review]
Actually, do we need the json/xml descriptors?
I'm not sure they're useful, but I'm sure they're costly to maintain.

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php (hunk 3 above)
Comment (ro0NL) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php]: i took it from here https://github.com/symfony/symfony/blob/e00d2f14fbaf947ad585216d811d5444e3217a36/src/Symfony/Component/Form/Console/Descriptor/TextDescriptor.php#L18 :) Basically we require vardumper to display complex values, but maybe we can do a fallback? maybe @yceruto has a suggestion

**Comment by ro0NL** [comment]
> Actually, do we need the json/xml descriptors?

not really, do you suggest to throw from any format other then text?

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php [hunk 5]
```diff
@@ -284,4 +292,35 @@ protected function sortServiceIds(array $serviceIds)
...
         return $serviceIds;
     }
+
+    private function getContainerEnvVars(ContainerBuilder $container): array
+    {
+        $getEnvReflection = new \ReflectionMethod($container, 'getEnv');
+        $getEnvReflection->setAccessible(true);
+        $envs = array();
+        foreach (array_keys($container->getEnvCounters()) as $env) {
+            $processor = 'string';
```
[the remark this answers is not in the document]
Comment (ro0NL) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/Descriptor.php]: it wont help much :) but in general i'd like to be explicit here. In the code it's `string` too

still hesitating about horizontal display vs. vertically. Ill try such a variant as well tomorrow.

**Comment by fabpot** [comment]
Throwing an exception sounds good to me.

**Comment by ro0NL** [comment]
@fabpot for 4.2? Do you suggest to keep the VarDumper dependency for now.. Form component already does it. Otherwise we need to squeeze #28898 as well

**Comment by ro0NL** [comment]
screenshots updated

**Comment by ro0NL** [comment]
i dropped JSON support as well, no real need for a machine format inspect/debug envs. They are bound to the current environment anyway.

File: src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/JsonDescriptor.php [hunk 6]
```diff
@@ -177,6 +177,14 @@ protected function describeContainerParameter($parameter, array $options = array
         $this->writeData(array($key => $parameter), $options);
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    protected function describeContainerEnvVars(array $envs, array $options = array())
+    {
+        throw new \LogicException('The JSON format is not supported.');
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/JsonDescriptor.php]: maybe `is not supported for debugging env vars.`?

Comment (chalasr) [reply src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/JsonDescriptor.php]: could use the component's `LogicException` also

**Comment by xabbuh** [comment]
@ro0NL Any chance to finish this one for 4.3? :)

**Comment by fabpot** [comment]
@ro0NL What needs to be done before merging?

**Comment by ro0NL** [comment]
@fabpot ready. Failures unrelated.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ro0NL.
